File: layers/L3_operational/architecture/dashboard/dashboard-server.py
#!/usr/bin/env python3
"""
HAL9 Integrated Dashboard Server
통합 대시보드 서버 - 모든 HAL9 서비스를 한 곳에서 모니터링
"""

import asyncio
import json
import os
import psutil
import time
from datetime import datetime
from http.server import HTTPServer, SimpleHTTPRequestHandler
from pathlib import Path
import threading
import requests
import random
from websocket import WebSocketServer

class DashboardServer:
    def __init__(self, port=8080):
        self.port = port
        self.start_time = time.time()
        self.metrics = {
            "system": {
                "cpu": 0,
                "memory": 0,
                "latency": 0,
                "uptime": 0
            },
            "neurons": {
                "active": 0,
                "layers": 0,
                "speed": 0,
                "grid": [False] * 100
            },
            "consciousness": {
                "phi": 0.0,
                "ratio": "1:1"
            },
            "games": {
                "active": 0,
                "winRate": 0
            },
            "performance": {
                "ops": 0,
                "avgResponse": 0,
                "fps": 0
            }
        }
        self.websocket_server = None
        self.services = {
            "api-main": "http://localhost:3456",
            "consciousness": "http://localhost:8765",
            "self-org": "http://localhost:8766"
        }
        
    def update_metrics(self):
        """시스템 메트릭 업데이트"""
        # 실제 시스템 정보
        self.metrics["system"]["cpu"] = psutil.cpu_percent(interval=0.1)
        self.metrics["system"]["memory"] = psutil.virtual_memory().percent
        self.metrics["system"]["uptime"] = int(time.time() - self.start_time)
        
        # 서비스 연결 상태 확인
        latencies = []
        for service, url in self.services.items():
            try:
                start = time.time()
                response = requests.get(url, timeout=1)
                latency = int((time.time() - start) * 1000)
                latencies.append(latency)
            except:
                latencies.append(100)  # 연결 실패시 100ms
        
        self.metrics["system"]["latency"] = sum(latencies) // len(latencies) if latencies else 0
        
        # HAL9 서비스에서 실제 데이터 가져오기
        try:
            # AI Genius Game API
            games_response = requests.get(f"{self.services['api-main']}/api/games", timeout=1)
            if games_response.status_code == 200:
                games = games_response.json()
                self.metrics["games"]["active"] = len(games)
        except:
            pass
        
        # 시뮬레이션 데이터 (실제 서비스 연동시 교체)
        self.metrics["neurons"]["active"] = random.randint(1000, 10000)
        self.metrics["neurons"]["layers"] = random.randint(2, 8)
        self.metrics["neurons"]["speed"] = round(random.uniform(1, 100), 2)
        self.metrics["neurons"]["grid"] = [random.random() > 0.7 for _ in range(100)]
        
        self.metrics["consciousness"]["phi"] = round(random.uniform(0, 2), 3)
        self.metrics["consciousness"]["ratio"] = f"{round(random.uniform(1, 10), 1)}:1"
        
        self.metrics["games"]["winRate"] = random.randint(60, 95)
        
        self.metrics["performance"]["ops"] = random.randint(10000000, 100000000)
        self.metrics["performance"]["avgResponse"] = round(random.uniform(10, 100), 1)
        self.metrics["performance"]["fps"] = random.randint(30, 120)
    
    def broadcast_metrics(self):
        """WebSocket으로 메트릭 브로드캐스트"""
        if self.websocket_server:
            message = json.dumps(self.metrics)
            self.websocket_server.send_message_to_all(message)
    
    def handle_websocket_message(self, client, server, message):
        """WebSocket 메시지 처리"""
        try:
            data = json.loads(message)
            action = data.get("action")
            
            if action == "reorganize":
                print("[Dashboard] Reorganizing neurons...")
                # 뉴런 재조직화 로직
                response = {"status": "success", "message": "Neurons reorganized"}
                
            elif action == "add_neurons":
                count = data.get("count", 10)
                print(f"[Dashboard] Adding {count} neurons...")
                # 뉴런 추가 로직
                response = {"status": "success", "added": count}
                
            else:
                response = {"status": "error", "message": "Unknown command"}
            
            server.send_message(client, json.dumps(response))
            
        except Exception as e:
            print(f"[Dashboard] Error handling message: {e}")
    
    def run_metrics_updater(self):
        """메트릭 업데이트 루프"""
        while True:
            self.update_metrics()
            self.broadcast_metrics()
            time.sleep(1)
    
    def start(self):
        """대시보드 서버 시작"""
        print(f"🌌 HAL9 Integrated Dashboard Server")
        print(f"=" * 50)
        
        # WebSocket 서버 시작
        self.websocket_server = websocket_server.WebsocketServer(9001)
        self.websocket_server.set_fn_message_received(self.handle_websocket_message)
        
        ws_thread = threading.Thread(target=self.websocket_server.run_forever)
        ws_thread.daemon = True
        ws_thread.start()
        print(f"✓ WebSocket server started on port 9001")
        
        # 메트릭 업데이터 시작
        metrics_thread = threading.Thread(target=self.run_metrics_updater)
        metrics_thread.daemon = True
        metrics_thread.start()
        print(f"✓ Metrics updater started")
        
        # HTTP 서버 시작
        class DashboardHandler(SimpleHTTPRequestHandler):
            def __init__(self, *args, dashboard_server=None, **kwargs):
                self.dashboard_server = dashboard_server
                super().__init__(*args, **kwargs)
            
            def do_GET(self):
                if self.path == "/" or self.path == "/dashboard":
                    self.serve_dashboard()
                elif self.path == "/api/dashboard/metrics":
                    self.serve_metrics()
                else:
                    super().do_GET()
            
            def serve_dashboard(self):
                dashboard_path = Path(__file__).parent / "integrated-dashboard.html"
                if dashboard_path.exists():
                    self.send_response(200)
                    self.send_header("Content-type", "text/html")
                    self.end_headers()
                    with open(dashboard_path, "rb") as f:
                        self.wfile.write(f.read())
                else:
                    self.send_error(404, "Dashboard not found")
            
            def serve_metrics(self):
                self.send_response(200)
                self.send_header("Content-type", "application/json")
                self.end_headers()
                metrics = self.server.dashboard_server.metrics
                self.wfile.write(json.dumps(metrics).encode())
            
            def log_message(self, format, *args):
                # 로그 억제
                pass
        
        # Handler에 dashboard_server 참조 추가
        handler = lambda *args, **kwargs: DashboardHandler(
            *args, 
            dashboard_server=self, 
            **kwargs
        )
        
        httpd = HTTPServer(("", self.port), handler)
        httpd.dashboard_server = self
        
        print(f"✓ HTTP server started on port {self.port}")
        print(f"\n🚀 Dashboard available at: http://localhost:{self.port}")
        print(f"📡 WebSocket endpoint: ws://localhost:9001")
        print(f"\nPress Ctrl+C to stop the server")
        print(f"=" * 50)
        
        try:
            httpd.serve_forever()
        except KeyboardInterrupt:
            print("\n\nShutting down dashboard server...")
            httpd.shutdown()

def main():
    """메인 함수"""
    # 환경 변수에서 포트 읽기
    port = int(os.environ.get("DASHBOARD_PORT", 8080))
    
    # 대시보드 서버 시작
    server = DashboardServer(port)
    server.start()

if __name__ == "__main__":
    main()

File: layers/L3_operational/architecture/dashboard/dashboard-server.rs
use axum::{
    extract::{ws::WebSocket, State, WebSocketUpgrade},
    response::{Html, IntoResponse},
    routing::{get, post},
    Json, Router,
};
use serde::{Deserialize, Serialize};
use std::{
    collections::HashMap,
    net::SocketAddr,
    sync::Arc,
    time::{Duration, Instant},
};
use tokio::{
    sync::{Mutex, RwLock},
    time::interval,
};
use tower_http::cors::CorsLayer;
use tracing::{info, warn};

#[derive(Clone)]
struct DashboardState {
    metrics: Arc<RwLock<SystemMetrics>>,
    connections: Arc<Mutex<HashMap<String, WebSocket>>>,
    start_time: Instant,
}

#[derive(Clone, Serialize, Default)]
struct SystemMetrics {
    system: SystemInfo,
    neurons: NeuronStatus,
    consciousness: ConsciousnessMetrics,
    games: GameStats,
    performance: PerformanceMetrics,
}

#[derive(Clone, Serialize, Default)]
struct SystemInfo {
    cpu: f32,
    memory: f32,
    latency: u32,
    uptime: u64,
}

#[derive(Clone, Serialize, Default)]
struct NeuronStatus {
    active: usize,
    layers: usize,
    speed: f32,
    grid: Vec<bool>,
}

#[derive(Clone, Serialize, Default)]
struct ConsciousnessMetrics {
    phi: f64,
    ratio: String,
}

#[derive(Clone, Serialize, Default)]
struct GameStats {
    active: usize,
    #[serde(rename = "winRate")]
    win_rate: u8,
}

#[derive(Clone, Serialize, Default)]
struct PerformanceMetrics {
    ops: u64,
    #[serde(rename = "avgResponse")]
    avg_response: f32,
    fps: u32,
}

#[derive(Deserialize)]
struct DashboardCommand {
    action: String,
    #[serde(default)]
    count: Option<usize>,
}

impl DashboardState {
    fn new() -> Self {
        Self {
            metrics: Arc::new(RwLock::new(SystemMetrics::default())),
            connections: Arc::new(Mutex::new(HashMap::new())),
            start_time: Instant::now(),
        }
    }

    async fn update_metrics(&self) {
        let mut metrics = self.metrics.write().await;
        
        // Update system info
        metrics.system.cpu = get_cpu_usage();
        metrics.system.memory = get_memory_usage();
        metrics.system.latency = measure_latency().await;
        metrics.system.uptime = self.start_time.elapsed().as_secs();
        
        // Update neuron status
        metrics.neurons.active = rand::random::<usize>() % 10000;
        metrics.neurons.layers = (rand::random::<usize>() % 6) + 2;
        metrics.neurons.speed = rand::random::<f32>() * 100.0;
        metrics.neurons.grid = (0..100).map(|_| rand::random::<bool>()).collect();
        
        // Update consciousness metrics
        metrics.consciousness.phi = rand::random::<f64>() * 2.0;
        metrics.consciousness.ratio = format!("{:.1}:1", rand::random::<f32>() * 10.0);
        
        // Update game stats
        metrics.games.active = rand::random::<usize>() % 10;
        metrics.games.win_rate = rand::random::<u8>() % 100;
        
        // Update performance
        metrics.performance.ops = rand::random::<u64>() % 100_000_000;
        metrics.performance.avg_response = rand::random::<f32>() * 100.0;
        metrics.performance.fps = rand::random::<u32>() % 120;
    }

    async fn broadcast_metrics(&self) {
        let metrics = self.metrics.read().await.clone();
        let message = serde_json::to_string(&metrics).unwrap();
        
        let mut connections = self.connections.lock().await;
        connections.retain(|id, ws| {
            // In real implementation, send through WebSocket
            // For now, just log
            info!("Broadcasting to connection {}: {}", id, message);
            true
        });
    }
}

async fn serve_dashboard() -> Html<&'static str> {
    Html(include_str!("integrated-dashboard.html"))
}

async fn websocket_handler(
    ws: WebSocketUpgrade,
    State(state): State<DashboardState>,
) -> impl IntoResponse {
    ws.on_upgrade(move |socket| handle_socket(socket, state))
}

async fn handle_socket(socket: WebSocket, state: DashboardState) {
    let id = uuid::Uuid::new_v4().to_string();
    info!("New dashboard connection: {}", id);
    
    // Store connection
    state.connections.lock().await.insert(id.clone(), socket);
    
    // Handle incoming messages
    // In real implementation, process commands from dashboard
}

async fn get_metrics(State(state): State<DashboardState>) -> Json<SystemMetrics> {
    Json(state.metrics.read().await.clone())
}

async fn handle_command(
    State(state): State<DashboardState>,
    Json(command): Json<DashboardCommand>,
) -> Json<serde_json::Value> {
    match command.action.as_str() {
        "reorganize" => {
            info!("Reorganizing neurons...");
            // Trigger neuron reorganization
            Json(serde_json::json!({ "status": "success" }))
        }
        "add_neurons" => {
            let count = command.count.unwrap_or(10);
            info!("Adding {} neurons...", count);
            // Add neurons to the system
            Json(serde_json::json!({ "status": "success", "added": count }))
        }
        _ => Json(serde_json::json!({ "status": "error", "message": "Unknown command" })),
    }
}

// Mock functions - replace with real implementations
fn get_cpu_usage() -> f32 {
    rand::random::<f32>() * 100.0
}

fn get_memory_usage() -> f32 {
    rand::random::<f32>() * 100.0
}

async fn measure_latency() -> u32 {
    rand::random::<u32>() % 100
}

pub async fn run_dashboard_server() {
    let state = DashboardState::new();

    // Spawn metrics update task
    let metrics_state = state.clone();
    tokio::spawn(async move {
        let mut interval = interval(Duration::from_secs(1));
        loop {
            interval.tick().await;
            metrics_state.update_metrics().await;
            metrics_state.broadcast_metrics().await;
        }
    });

    // Build router
    let app = Router::new()
        .route("/", get(serve_dashboard))
        .route("/dashboard", get(serve_dashboard))
        .route("/ws/dashboard", get(websocket_handler))
        .route("/api/dashboard/metrics", get(get_metrics))
        .route("/api/dashboard/command", post(handle_command))
        .layer(CorsLayer::permissive())
        .with_state(state);

    let addr = SocketAddr::from(([0, 0, 0, 0], 8080));
    info!("Dashboard server running on http://{}", addr);

    axum::Server::bind(&addr)
        .serve(app.into_make_service())
        .await
        .unwrap();
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn test_dashboard_state() {
        let state = DashboardState::new();
        state.update_metrics().await;
        
        let metrics = state.metrics.read().await;
        assert!(metrics.system.cpu >= 0.0 && metrics.system.cpu <= 100.0);
        assert!(metrics.neurons.active > 0);
    }
}